### ISSUE TITLE
PositiveLongAttributeModel and PositiveDoubleAttributeModel getInvalidValueCause method is sensitive to null value

### ISSUE THREAD

**Issue by cyril2maq** [eclipse-leshan/leshan#1542]
https://github.com/eclipse-leshan/leshan/blob/2afadfcc35f8ab78d471eee8f95c6213607e8ad5/leshan-core/src/main/java/org/eclipse/leshan/core/link/lwm2m/attributes/PositiveLongAttributeModel.java#L34

I guess 'null' value is supposed to be a valid use case to 'unset' an attribute; with this is not possible anymore with these implementations.

Sample test that was valid in M13 and is not anymore (it throws a NullPointerException) :

 ```java
    @Test
    void create_pmin_attribute_with_null_value() throws Exception {
        LwM2mAttribute<Long> pmin = LwM2mAttributes.create(LwM2mAttributes.MINIMUM_PERIOD, null);
        assertThat(pmin.getValue()).isNull();
    }
```

But may be there is another way to create such 'unset' attribute ?

**Comment by sbernard31** [comment]
Thx testing and reporting this. :pray: 

I look at this a little and for now, I'm not sure if this is a bug :thinking: 

The "right" way to create an attribute without value is : `LwM2mAttributes.create(LwM2mAttributes.MINIMUM_PERIOD);` but I'm not sure what should be the right behavior for :`LwM2mAttributes.create(LwM2mAttributes.MINIMUM_PERIOD, null);` (I mean should we forbid `null` or allow it like if `LwM2mAttributes.create(LwM2mAttributeModel<T>)` method was called ?

Do you face this only in this test OR also in other more concrete case ?

**Comment by sbernard31** [comment]
I looked at Leshan code and we never call `LwM2mAttributes.create(model, value)` with a `null` value.

We also have some test which check that parsing "Notification" without value are OK : 
https://github.com/eclipse-leshan/leshan/blob/2afadfcc35f8ab78d471eee8f95c6213607e8ad5/leshan-core/src/test/java/org/eclipse/leshan/core/link/lwm2m/attributes/DefaultLwM2mAttributeParserTest.java#L34-L57

So I think this is more about clarifying current API behavior.
Knowing that only some Attribute can be created empty, (I think only attribute of class "notification") maybe better to be strict and adapt code in this way : 
 - `LwM2mAttributes.create(model)` must be the only way to create attribute without value.
 -  only "notification" attribute can used with `LwM2mAttributes.create(model)`
 - `LwM2mAttributes.create(model, value)` any call of this method with `null` should raise an exception. 
 - javadoc should be adapted
 
 Let me know if you have any opinion about it.
 I think this can be done after the 2.0.0-M14 release because this is not a critical behavior issue. 
 
 Let me know your opinion about that. :pray:

**Comment by cyril2maq** [comment]
It was in a test use case, where we wanted to unset a value.
But I missed the other method call (has we use it in a generic way - with null and non null parameter).
This alternative works fine for us; we just have to null check beforehand and then call the right method. So we can handle it for our M14 update.

But it may seem inconsistent with DoubleAttributeModel for instance, where LwM2mAttributes.create(LwM2mAttributes.LESSER_THAN, null) is working fine.

It is a breaking change in the behaviour, maybe add this in the release note ?

**Comment by cyril2maq** [comment]
It would be a great solution 👍

**Comment by sbernard31** [comment]
I push a fix about it in `master` : d9a0688.

Some tests are disabled because they don't pass. It should be fixed post 2.0.0-M14 release.

**Comment by sbernard31** [comment]
So now the **2.0.0-M14** is released, we should decide the expected behavior for LwM2mAttributes.create(model, value)  API. 

My first thought was to be strict and never accept `null` as value for `LwM2mAttributes.create(model, value)`

But reading https://github.com/eclipse-leshan/leshan/issues/1542#issuecomment-1822949200, I understand that this could maybe mode convenient to be able to call LwM2mAttributes.create(model, value) with null value when possible.

@cyril2maq could you confirm that you think that 2nd solution sounds better to you ?

**Comment by cyril2maq** [comment]
Indeed, it would be more convenient for our use case.
Usually I go for strictness, but here, as value=null is semantically correct, and is equivalent to .create(model) I do think it seems cumbersome to be strict.
So, allowing value=null sounds better to me.

**Comment by sbernard31** [comment]
Ok so I will go in that way.

**Comment by sbernard31** [comment]
I tried to fix this more properly at #1551.

@cyril2maq, Let me know if you want to review/look at it before I integrate it in `master`. 
(no problem if you didn't have time to do that)

**Comment by cyril2maq** [comment]
Sorry, I will not be able to review it. 
Nevertheless, I did pull this commit, and I confirm that our tests are 🟢 with this commit (tests which previously failed due to the null value).

**Comment by sbernard31** [comment]
@cyril2maq thx for testing :pray:.

I integrated #1551 in `master` will be available in next release `2.0.0-M15`.

I close this issue.
